### ISSUE TITLE
Projecting single files from volumes into a container's FS

### ISSUE THREAD

**Issue by pmorie** [kubernetes/kubernetes#19764]
This issue is to explore the possibility of allowing users to make projections of individual files from volumes into a container.  Example use-case is where the user wants to place volume content into an existing directory without mounting over the existing contents of that directory.  Example:  say you want to project a file into `/etc/ssl/certs` without making the other files in that directory inaccessible.  

@mfojtik This is your use-case -- feel free to elaborate.

@thockin @smarterclayton @bgrant0607 @saad-ali @kubernetes/rh-storage

**Comment by pmorie** [comment]
After some thought, I realize that if we do this, there's really no need for #19762

**Comment by thockin** [comment]
So the biggest issue with this is that updating a file atomically across a
namespace is practically impossible.  You have to enter the target
namespace to really do it right.

I'd love to see this be possible, so it is probably worth exploring, but I
think it HAS to support atomic updates properly if updates are supported at
all.

**Comment by mfojtik** [comment]
@pmorie thanks! another use-case we have is to add files like `.npmrc` into `$HOME` or to `/etc`. This file contains internal URL to nodejs mirror and credentials (user+pass). This file is needed during the build time and not needed in output image (so having a way to 'unplug' the file from the contains would be nice to have as well ;-)

**Comment by mfojtik** [comment]
@thockin I think the update to secret content should trigger restart of the container where the secret is projected to.

**Comment by thockin** [comment]
This is a point of debate for secrets.  DownwardAPI already allows dynamic
refresh, and config map is not done yet.

**Comment by bgrant0607** [comment]
Some applications expect configuration files directly in /etc. This would be useful for such cases.

Yes, we need atomic updates for some cases.

**Comment by pmorie** [comment]
I think this is also solvable with pre-start hooks

**Comment by thockin** [comment]
Apps that NEED it in /etc/foo.conf can provide an image that symlinks
/etc/foo.conf -> /vol/etc/foo.conf.

I'm not against this feature, I think it would be great, I just don't think
the implementation (at least the one I know how to do) is justified, when a
relatively easy workaround is present.

**Comment by pmorie** [comment]
@thockin Agree, if you really want this, you can have it without a kube feature.  

@mfojtik How do you feel about setting up symlinks in the container entrypoint, does that get you any closer to where you want to be?

**Comment by smarterclayton** [comment]
Changing the image is actually a fairly involved process - the statement we
make in other places is that adaptation of existing images to run is useful
and justifies some additional complexity on our end to make easy.
Individual file projection is useful in some cases, and less useful in
others.  As someone who has done the symlinking thing, it's error prone,
annoying to do, and you end up doing it in a lot of images (if you have
more than 2 you care about).  So there's some value in having a bit more
complexity.  I don't know that it's required *right now*

**Comment by thockin** [comment]
I agree this would be a great feature, I don't think it is "a bit more
complexity", though.  If someone wants to prototype it, I'm happy to brain
dump (or just go look at https://github.com/docker/docker/issues/17190)

**Comment by erictune** [comment]
A customer had a narrower problem which might have a more direct solution -- no sure if it is worthwhile.
The customer wanted to have secrets A and B, each with one key, say "file-a" and "file-b".  The customer wanted to mount them at "/etc/secrets/file-a" and "/etc/secrets/file-b".  I suppose we could recognize this specific case at kubelet and merge the keys into a single combined volume prior to mounting.  But, there are probably enough sharp corners that I'm not sure this is worthwhile.  Also, explaining what is and isn't supported would get harder.

**Comment by thockin** [comment]
Options I see:

a) do the really complicated process for updating a file-projection (enter
target namespace...)

b) document that single-file projections are not live-updatable

c) document the (user-driven) workaround of making an image with
/etc/secret-a be a symlink -> /mnt/secrets/a

d) unpack the volume image ourselves and reconstruct the parent dir and
mount the whole dir (likely to break in other ways, e.g. docker commit)

e) convince upstream kernels to support bind-mounting symlinks as symlinks
(without dereference)  This is a good answer if we can take a long view
(though it may be a nonstarter, I don't know kernel VFS that well)

**Comment by thockin** [comment]
It is not sufficient or workable to build the image with the file being a symlink to an empty dir, and let secret or configmap mount into the dir?  Not that I don't want this feature, but I want to see the f we can unblock people.

**Comment by eparis** [comment]
I know that @derekwaynecarr also had this problem. openshift expects certs and config files to live next to each other in the same directory. Yes, his problem is 'solvable'/'work-arround-able' by creating a custom image with symlinks. But custom images come with their own pain when the original image is updated.

**Comment by smarterclayton** [comment]
It's fairly common to have secrets and config mixed (off hand I think
most Java app servers, some webservers, and lots of normal software
have the same directory).  I suspect most people would just assume it
would work (that things would end up into the same location).  I
suspect they'd be willing to accept one or the other "winning".

It's not the end of the world, but I suspect we'll have lots of users
using config map and secret hit this, then try to do an ugly
workaround.

**Comment by pmorie** [comment]
@smarterclayton @eparis @thockin Why not have an all-in-one volume for this for the time being?

**Comment by thockin** [comment]
The all-in-one was really ugly, I thought.  I guess I am not really against it, it just feels clumsier than it should be.  any other answer I have come up with is uglier, though.  Still doesn't solve the as-a-file case.

**Comment by aanm** [comment]
Continuing from #23237, the `hostPath` http://kubernetes.io/docs/user-guide/volumes/#hostpath exists, so isn't to difficult to merge both? I've looked at the code and I know that would take a lot of work than just magically merge it.

**Comment by thockin** [comment]
Read the rest of this issue.  Doing it is not that hard, as long as we give
up on the ability to ever dynamically update contents, which we CAN do with
directories.

**Comment by dshulyak** [comment]
Shouldn't this be closed by https://github.com/kubernetes/kubernetes/pull/22575?

**Comment by thockin** [comment]
I am not sure this is going to be fixed any better than it is (it does work, now, but no updates) in the near term, so yeah, i'm gonna close this.

Thanks!

**Comment by pmorie** [comment]
@dshulyak #22575 doesn't fix the problem because the kernel bind-mounts inodes, not symlinks.  So, you can use subPath to get the _initial value_ of a file in this manner, but updates won't work correctly with subPath.

**Comment by thockin** [comment]
It might be a fun project to negotiate a kernel change that does what we
want here, but that's a long-running effort.  still, I would endorse
someone earning their kernel-hacking wings on it.

**Comment by ienliven** [comment]
I ran into this problem, had to use Pod lifecycle hook PostStart to link the file from configmap into desired location. 

<pre>
    ...
    spec:
      containers:
        - image: wordpress
          lifecycle:
            postStart:
              exec:
                command:
                  - ln
                  - -s
                  - /usr/local/etc/php-config/uploads.ini
                  - /usr/local/etc/php/conf.d/uploads.ini
          name: wordpress-site
          ...
</pre>

**Comment by davey-dev** [comment]
@ienliven Nice workaround. Might have to "ln -fs" if the file already exists.  PostStart is NOT guaranteed to execute before the container entrypoint though :(

**Comment by james-woods** [comment]
I have some kind of the same issue. I am trying to build a pod that has access to Google Cloud Services and I am mounting my secrets into it so that my service account can talk via the standard google cloud toolchain like `bq`. Symlinking the credentials into `/.config/gcloud/` breaks since

`oauth2client.contrib.locked_file.CredentialsFileSymbolicLinkError: File: /.config/gcloud/credentials is a symbolic link.`

Hacking around this by copying via PostStart hook helps.

**Comment by deitch** [comment]
> it does work, now, but no updates

Does it? I could not figure out how. If I have an app that expects a config file in `/app.conf` or `/app/conf/app.conf`, what syntax do I use to mount the configmap as the single file `app.conf` without mounting over the entire `/` or `/app/conf/`?

**Comment by deitch** [comment]
Well, it looks like I figured it out here https://github.com/dshulyak/kubernetes.github.io/commit/d58ba7b075bb4848349a2c920caaa08ff3773d70

but that was a lot harder to find than it should have been.

**Comment by bryanlarsen** [comment]
getting @deitch's example into the ConfigMap docs would be nice.

**Comment by bitmage** [comment]
I'm using [ployst/nginx-ssl-proxy](https://github.com/ployst/docker-nginx-ssl-proxy) which looks for all its secrets in `/etc/secrets`.  This worked fine when I just had secrets coming from one place (the letsencrypt container), but now that I want to add basic auth, things get complicated.  I first tried:

```yaml
command: ["/bin/sh", "-c", "ln -s /etc/auth/htpasswd /etc/secrets/htpasswd && ./start.sh"
```

I wasn't able to mount the cert secrets to `/etc/secrets` and then just symlink the auth secret.  The symlink command would run successfully, but when the container started the `htpasswd` file would be gone.  If I then `exec -it [pod] /bin/bash` I can manually run the symlink command and the file persists.  This makes me think that maybe a volume refresh is being triggered during the container finalization lifecycle.  Speculation though - there's really no evidence of what happened.

I ended up mounting both of my secrets to separate locations, then symlinking all the files I need:

```yaml
command: ["/bin/sh", "-c", "mkdir -p /etc/secrets && \
                            ln -s /etc/certs/dhparam /etc/secrets/dhparam && \
                            ln -s /etc/certs/proxycert /etc/secrets/proxycert && \
                            ln -s /etc/certs/proxykey /etc/secrets/proxykey && \
                            ln -s /etc/auth/htpasswd /etc/secrets/htpasswd && \
                            ./start.sh"
```

It's an unfortunate amount of jumping through hoops though.

**Comment by deitch** [comment]
> It's an unfortunate amount of jumping through hoops though

Bit of a theme....

**Comment by thockin** [comment]
@bitmage if you can boil down a repro, we can look into it.  Naively, I would expect what you described to work...

@jingxu97 @saad-ali

**Comment by debianmaster** [comment]
facing same issue, any officially suggested workarounds for this problem?

**Comment by debianmaster** [comment]
nm found it!

**Comment by fearphage** [comment]
@debianmaster Care to share the official response that you found?

**Comment by schmitzhermes** [comment]
Just spend several minutes searching for a solution, found it here: https://github.com/kubernetes/kubernetes/issues/23748#issuecomment-230390309

you need to supply the **absolute path including the filename** in the `mountPath` and the filename again in `subPath`. The filename (obviously) needs to match the key in your config map.

**Comment by debianmaster** [comment]
@schmitzhermes   @fearphage sorry for not getting back.    you have the right solution @schmitzhermes    it wont work from cli though if you have more than 1 file to mount in the same directory.  i.e.  only works for 1 file

**Comment by ketchoop** [comment]
@schmitzhermes unfortunately, as I can see, configmap can't be updated when subPath is defined :(

**Comment by smarterclayton** [comment]
I cry a little bit every time I hit this limitation.  It's definitely one
of my top 10 usability issues for any complex pod (since a lot of Unix
software uses `/etc/` and we can't project into `/etc` without a hammer or
hacks.

**Comment by thockin** [comment]
@pmorie and I had sketched out a "merge" volume, but we'd need an ower

**Comment by funlake** [comment]
+1 cry every time,any plans for single file volume mapping?

**Comment by thockin** [comment]
There are more feature requests than people to implement. ;(

**Comment by Chrislevi** [comment]
Any news? hit the same wall

**Comment by SleepyBrett** [comment]
This becomes an even bigger issue when you widely enable restricted psps that don't allow for rootfs writes.... which we did last week.

**Comment by thockin** [comment]
I'd love to see this get worked on, but it needs an owner.  I'm happy to reopen if someone wants to tackle it..

**Comment by ffledgling** [comment]
@thockin what would implementing this take? I saw you mention in one of the related issues that it's technically hard to do.
